### DISCUSSION TITLE
Related Recipes

### DISCUSSION THREAD

**Discussion by BCNelson** [mealie-recipes/mealie#1554] [Feature Request]
**Before submitting this feature request I have**
- [x] Reviewed the documentation and verified that my feature request isn't already a feature.
- [x] Reviewed the existing feature requests and my feature has not been requested.
- [x] Checked the [tasks taged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**
Recipe often have many variations of the same recipe 
i.e
- Chicken Katsu vs Pork Katsu 
- Chicken Alfredo vs Shrimp Alfredo
- Vegan vs Non vegan

**(Optional): Suggest A Solution**
There are two options in my mind
1)  Deep Ingratiation
2) Minimum Possible
2b) Minimum with some flair

Starting with the minimum possible changes as I think that is is more likely what would be implemented
2)  Create a Many to Many Relationship between recipes and show them on the recipe detail page
2b) Same as 2 adding types to these relationship that allow for the type of link to be noted possible links off of the top of my head include
- Variations
- Good Pairings
- Ingredient Recipe  (A recipe who's output is used in this recipe i.e  chicken broth or bread)
1) Deep integration would include having a base recipe and then layering the difference of the variation on top of that. Liking this would require a substantial rework and be very intrusive as well as complicating the UI. Although it could be done by adding a variation table that links to the parent recipe and allow hiding of the parent form default search results as it is essentially an abstract class

If there are multiple solutions, please present each one separately. Save comparisons for the very end.)

**Additional Information**
- [x] If the feature is accepted I'm willing to submit a PR to provide this feature
I'm willing to open a PR for option 2. Options 1 I'm willing to help but i think that is beyond what I have time for

**Comment by BCNelson** [comment]
Looking to it more it looks like there is not an obviously best way to implement many to many self referential relationships in sqlalchemy. I don't have any experience with sqlalchemy so could use some help with the design for someone who does.

**Reply by BCNelson** [reply]
Okay so looking into it for what I can tell there are two good options.
1) Have a many to many relationship that is self referential like this.
```py
class RecipeLinksModel(SqlAlchemyBase, BaseMixins):
    __tablename__ = "recipe_links"
    left_id = sa.Column(sa.ForeignKey("left.id"), primary_key=True)
    right_id = sa.Column(sa.ForeignKey("right.id"), primary_key=True)
    type = sa.Column(sa.String(50))

class RecipeModel(SqlAlchemyBase, BaseMixins):
    __tablename__ = "recipes"

    id = sa.Column(GUID, primary_key=True, default=GUID.generate)
    links = db.relationship('RecipeLinksModel',
        secondary=RecipeLinksModel,
        primaryjoin=(RecipeLinksModel.c.left_id)==id),
        secondaryjoin=(RecipeLinksModel.c.right_id==id),
        backref = db.backref('links', lazy = 'dynamic'), 
        lazy = 'dynamic')
    def list_links(self):
        links_union = db.select([
                        RecipeLinksModel.c.left_id, 
                        RecipeLinksModel.c.right_id
                        ]).union(
                            db.select([
                                RecipeLinksModel.c.left_id, 
                                RecipeLinksModel.c.right_id]
                            )
                    ).alias()
        RecipeModel.all_links = db.relationship('RecipeModel',
                       secondary=links_union,
                       primaryjoin=RecipeModel.id==RecipeLinksModel.c.left_id,
                       secondaryjoin=RecipeModel.id==RecipeLinksModel.c.right_id,
                       viewonly=True) 
        return self.all_links
```
2) Is changing this to a directional relation ship so can skip the checks and the union. This would probably mean that the client has more work to do. Potently asking if the user waits it to create both sides of the link.

**Comment by michael-genson** [comment]
You can link to other recipes using HTML tags, e.g.

`<a href="https://my-mealie-site.com/g/home/r/my-other-recipe">Non-vegan-version</a>`

Similar discussion here: https://github.com/mealie-recipes/mealie/discussions/1761
